Import export

jbradberry opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Implement `import` and `export` commands in the new AWX CLI tool.

related #5146 

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - CLI

##### AWX VERSION
```
awx: 13.0.0
```

jbradberry committed:
Add skeleton of a custom 'export' resource

This only takes a flag for users at the moment, and does nothing in
particular with it.

jbradberry committed:
Normalize the requested resource into a pk

using the pk_or_name helper function.  Also, authenticate to the API.

jbradberry committed:
Actually query the API for the user or users requested

jbradberry committed:
Limit export output to only those fields needed to create the resource

jbradberry committed:
Add basic import command

jbradberry committed:
Add organizations as a resource type to export

jbradberry committed:
Use a list of exportable resources to build up the parser

Also, handle an implicit export of everything.

jbradberry committed:
Split out separate methods for the OPTIONS call and massaging each asset dict

jbradberry committed:
Resolve the natural keys of assets we've already collected

jbradberry committed:
Start to record the role membership

but not really, since it still isn't fully capturing the role content_object

jbradberry committed:
Fold the other methods doing API calls into get_assets()

jbradberry committed:
Make more use of the functionality of Page in get_assets

and related methods.  Also, rename get_resource_options -> get_options
and enhance_asset -> serialize_asset.

jbradberry committed:
Modify get_natural_key to work on Pages

jbradberry committed:
Make serialize_asset use Pages to a greater extent

jbradberry committed:
Extract out get_natural_key as a standalone function

jbradberry committed:
Begin changing Import to work with the structure created by Export

jbradberry committed:
Add the calculated natural key to the export data for each asset

jbradberry committed:
Begin to deal with existing matching assets

by using a registry allowing the lookup of Page objects by (frozen)
natural keys.

jbradberry committed:
Get and assign the related objects

jbradberry committed:
Fix a Python 2 syntax error

it doesn't like multiple **kwargs, apparently.

elyezer commented on awxkit/awxkit/cli/resource.py:
```diff
@@ -122,6 +175,188 @@ def handle(self, client, parser):
+        for resource in EXPORTABLE_RESOURCES:
+            # This parsing pattern will result in 3 different possible outcomes:
+            # 1) the resource flag is not used at all, which will result in the attr being None
+            # 2) the resource flag is used with no argument, which will result in the attr being ''
```
Does this mean that it is supposed to export everything?

jbradberry replied:
Everything of that type, yes.  E.g. `--users` by itself will export all users.

elyezer commented on awxkit/awxkit/cli/resource.py:
```diff
@@ -122,6 +175,188 @@ def handle(self, client, parser):
+            # This parsing pattern will result in 3 different possible outcomes:
+            # 1) the resource flag is not used at all, which will result in the attr being None
+            # 2) the resource flag is used with no argument, which will result in the attr being ''
+            # 3) the resource flag is used with an argument, and the attr will be that argument's value
```
Does this mean that it is supposed to export only the specified resource name?

jbradberry replied:
Yes.  `--users admin` should export only the user with username 'admin'.

jbradberry committed:
Only use a default password if an import user doesn't already exist

jbradberry committed:
Sort the asset groups to be imported by their dependency relationships

jbradberry committed:
Temporarily disable export of resources that may not work yet

jbradberry committed:
Filter out managed credential types

since we cannot patch them upon import.

jbradberry committed:
Create new Import.dependent_resources method

that yields up resource names, even ones that aren't explicitly in the
import data, in dependency order.

jbradberry committed:
Fix a couple of flaws

- The dependency ordering may spit out page types that weren't in the import data set.  Make sure to be able to map those to resources anyway.
- freeze() needs to be able to deal with nullable foreign keys.

jbradberry committed:
Enable credential export

AlanCoding commented on awxkit/awxkit/cli/resource.py:
```diff
@@ -44,6 +49,56 @@
+    'workflow_job_template': ('organization', 'name'),
+
+    # related resources
+    'role': ('name', ':content_object'),
```
Could you produce these from the named URL?

jbradberry replied:
No, I don't think so.  I'm deliberately packing all of the (potentially nested) information needed to reconstruct the link into the data structure, rather than relying on ad hoc program logic to figure it out from minimal information.  The results in the export look something like:

```json
    {
         "type": "role",
         "name": "Admin",
         "content_object": {
              "type": "credential",
              "organization": null,
              "name": "Demo Credential",
              "credential_type": {
                   "type": "credential_type",
                   "name": "Machine",
                   "kind": "ssh"
              }
         }
    },
```

AlanCoding commented on awxkit/awxkit/cli/resource.py:
```diff
@@ -44,6 +49,56 @@
+    'notification_template': ('organization', 'name'),
+    'project': ('organization', 'name'),
+    'inventory': ('organization', 'name'),
+    'job_template': ('name',),
```
I changed this, it may be `('organization', 'name')`, read-oriented, but `('project.organization', 'name')` in another write-oriented manner. I hope I did not screw everything up.

jbradberry replied:
Not a big deal, your work is on my radar.  I'm going to have to come up with a fallback scheme for missing identifiers anyway, not just for this but also the workflow template nodes.

AlanCoding replied:
> I'm going to have to come up with a fallback scheme for missing identifiers anyway

I feel like I might be working on something very similar with the AWX collection right now. Depending on your exact meaning. But that's in a totally different place, with significant differences in how the parameters are received.

jbradberry committed:
Deal somewhat reasonably with missing fields as part of the natural key

- JobTemplate.organization has recently been added, we need to support
  with and without
- WorkflowJobTemplateNode is shortly going to get an identifier field,
  and we will need to support both with and without

jbradberry committed:
Enable notification templates

jbradberry committed:
Move the export logic onto methods on the ApiV2 class

making it easier to invoke programmatically.

jbradberry committed:
Move the import logic into methods on the ApiV2 class

jbradberry committed:
Enable the export of several relation field types

jbradberry committed:
Enable the remaining top-level exportable resources

and rearrange the import/export methods a bit.

jbradberry committed:
Deal with lack of permissions

jbradberry committed:
Fix a bug with inventory.variables

Using Page[key] instead of Page.json[key] causes inner json blob
strings to automatically get parsed, which is not what we want with
this field.

jbradberry committed:
Export full related objects under some conditions

jbradberry committed:
Enable schedules

jbradberry committed:
Remove the natural key lookup dict and make each Page responsible instead

jbradberry committed:
Tentatively enable inventory sources

jbradberry committed:
If attempting to link to a resource that is not yet complete, wait

jbradberry committed:
Change the NoNaturalKey exception to no longer derive from Common

which seems to be entirely use for response exceptions.  Maybe rename
Common?

jbradberry committed:
Handle some more bad cases when doing OPTIONS calls

- deprecated endpoints
- read-only endpoints
- insufficient privileges

The latter case currently just drops it on the floor, but ought to do
something better.

jbradberry committed:
Remove the NoNaturalKey error

It's too awkward, and it makes more sense to return None instead.

jbradberry committed:
Drop objects that cannot be read or do not have a natural key

don't fail hard.

jbradberry committed:
Fall back to parsing the OPTIONS description

to determine the needed fields for constructing an object.

jbradberry committed:
Enable credential relations

but only when it is in an attach/detach list view.

jbradberry committed:
Enable interconnections between WFJT Nodes

jbradberry committed:
Fix the _create_assets method to use _get_options

jbradberry committed:
Hook up creation and attachment of related objects

jbradberry committed:
Deal with unreadable mandatory foreign keys

by dropping the parent object.  Also, clarify some of the warning log
messages.

jbradberry committed:
Remove the $encrypted$ placeholders from export values

jbradberry committed:
Only wait for Project updates

not other kinds of objects that have a status.

jbradberry committed:
Deal with relations that we cannot resolve

such as due to a lack of permissions.  If there is a foreign key to
something where we don't have sufficient read permissions, we now drop
the parent object from the export.

jbradberry committed:
Attempt to deal with foreign keys that get mislabeled as type integer

such as through the use of DeprecatedCredentialField.

jbradberry committed:
Split _get_options into two pieces

jbradberry committed:
Introduce a new PageCache object

and split out get_post_fields into its own utility function.

jbradberry committed:
Use the new PageCache to store and reuse Page.get results

jbradberry committed:
Avoid doing an OPTIONS call unless we know it is a related type we export

jbradberry committed:
Change the method that exports list views to take a PageList or TentativePage

This will allow the related m2m views to also use this method, with a
bit of effort.  Also, remove the use of pk_or_name in favor of a new
method that reduces the number of api calls.

jbradberry committed:
Rename some things for consistency

jbradberry committed:
Reuse _import_list for import of related full objects

e.g. WFJT Nodes.  Also rename _import_list from _create_assets.

jbradberry committed:
Log an error and continue when a related object is not found

jbradberry committed:
Make more use of the PageCache for imports

jbradberry committed:
Redo waiting until Project updates are complete

jbradberry committed:
Deal with exports involving foreign keys where you don't have permission

jbradberry committed:
Make sure we have a cached version of existing objects

even if it isn't an object being directly imported.  We might need it
for relations of things that are being imported.

jbradberry committed:
Recursively queue up related assets to be created and/or assigned

jbradberry committed:
Fix WFJT nodes endpoint to return node pages when posting

not the node list page.

chrismeyersfsu commented on awxkit/awxkit/api/pages/api.py:
```diff
@@ -13,7 +55,249 @@ class Api(base.Base):
+
+    def _export(self, _page, post_fields):
+        # Drop any (credential_type) assets that are being managed by the Tower instance.
+        if _page.json.get('managed_by_tower'):
```
Could this check go in a more credential-specific location?

jbradberry replied:
I've wanted for some time now to try to refactor all of this so that Page and PageList get export methods and this all isn't just piled up here on the ApiV2 class.  Each of my attempts so far has been a failure, though.  The export of a single object turns out to be annoyingly context specific.

chrismeyersfsu commented on awxkit/awxkit/api/pages/api.py:
```diff
@@ -13,7 +55,249 @@ class Api(base.Base):
+            if key not in _page.related:
+                continue
+
+            rel_endpoint = self._cache.get_page(_page.related[key])
```
explicit over implicit cache usage :+1:

chrismeyersfsu commented on awxkit/awxkit/api/pages/api.py:
```diff
@@ -13,7 +55,249 @@ class Api(base.Base):
+            rel_endpoint = self._cache.get_page(_page.related[key])
+            if rel_endpoint is None:  # This foreign key is unreadable
+                if post_fields[key].get('required'):
+                    log.error("Foreign key export failed: %s", _page.related[key])
```
Would be good to include the original record/url in the log message that led you down to the related url i.e. `_page.endpoint` since we hard bail via `return None`

jbradberry replied:
Yeah, I definitely need to clean up and improve my logging on this feature.

chrismeyersfsu commented on awxkit/awxkit/api/pages/api.py:
```diff
@@ -13,7 +55,249 @@ class Api(base.Base):
+                continue
+            natural_key = rel_endpoint.get_natural_key(self._cache)
+            if natural_key is None:
+                return None  # This foreign key has unresolvable dependencies
```
Include a `logg.error`

chrismeyersfsu commented on awxkit/awxkit/cli/resource.py:
```diff
@@ -123,6 +126,51 @@ def handle(self, client, parser):
+            parser.print_help()
+            raise SystemExit()
+
+        data = json.load(sys.stdin)
```
First time user interface feedback. The first time I ran `awx import` I suspected my awx server was going very slow. `awx` client was actually just waiting for stdin. That was unexpected. Maybe expect a filename by default and allow users to specify `--stdin` as the non-default.

chrismeyersfsu replied:
Similarly, the default for `export` should be an expected filename and `--stdout` as a non-default. In this case, it took a while to run through my AWX instance exporting. The output was then printed to STDOUT and I regretted not adding a `> /tmp/export.json`

jbradberry replied:
I agree, file-based I/O ought to be available.  Output has implications for the other subcommands, though, since I'm not doing anything specific for that.

ghjm replied:
I agree the current behavior is a bit unexpected and weird when you approach it for the first time.  Maybe check `sys.stdout.isatty()` early in the process and print a warning and exit if it's true.  If you think people might really want to do this, you could have a command line flag like `--no-tty-check` that suppresses the check.  And the converse for stdin.

chrismeyersfsu reviewed (dismissed):
A quick first past review. Going to explore using `awx import` and `awx export`

jbradberry committed:
Add more logging

jbradberry committed:
Return the changed status of imports

jbradberry commented:
regate

elyezer commented on awxkit/awxkit/api/pages/api.py:
```diff
@@ -1,6 +1,48 @@
+    'notification_templates',
+    'projects',
+    'inventory',
+    'inventory_sources',
```
Trying to export an inventory source I am seeing the following on the output:

```
This endpoint is deprecated: /api/v2/inventory_scripts/1/
Foreign key 'source_script' export failed for object /api/v2/inventory_sources/24/, setting to null
```

That means that the export functionality is partially exporting inventory sources that are created with a custom inventory script. When importing then, it will bring the information but not the script so it won't be able to sync that inventory source after the import process.

All that said, is this an intended behavior? Should we avoid export custom scripts inventory sources entirely? Should we provide a better message saying that the script won't be imported and that it would require some manual work?

jbradberry replied:
noting here for posterity: it turns out that the default behavior of the `cli` fixture in tower-qa is to lump together stdout and stderr.  This is not what we want to do for these tests, since as noted it will break json deserialization if there are _any_ log messages.

elyezer replied:
Tests will be updated after confirming that it is fine to not fully support inventory sources with inventory scripts.

The point is that the current functionality will export partial inventory sources when they have an inventory script on its current state. And those won't be able to sync after being imported because it will be missing the inventory script that was used to sync previously.

jbradberry committed:
Prioritize membership roles

since certain role grants will not be accepted by the api unless the user or team is part of the correct organization.

jbradberry committed:
Continue after failures to grant roles

shanemcd commented:
Can we add something under `docs/` for this? Maybe an `import_export.md`?

jbradberry committed:
First cut at examples of the import and export commands

jbradberry commented:
> Can we add something under `docs/` for this? Maybe an `import_export.md`?

@shanemcd How about this?  https://github.com/ansible/awx/blob/393369adec482f73e963b73832e7e3ff8b689542/awxkit/awxkit/cli/docs/source/examples.rst#importexport

ghjm commented:
FWIW, I ran this and it worked fine for all the things I tried.

elyezer commented:
I can't get the import to match the proper credentials owners, this is why the way that the import links a credential to users/organizations/teams is via the roles. The problem with that is that one can pass an owner to when creating a credential via the api (https://github.com/ansible/awx/blob/1321d298eee6ab9120f80037bd936ab25bc224d1/awx/api/serializers.py#L2619-L2633). If a team is passed, it will set some permissions to the credential (https://github.com/ansible/awx/blob/1321d298eee6ab9120f80037bd936ab25bc224d1/awx/api/serializers.py#L2666-L2667) but going to the team's roles it only has the `use` for that given credential but checking the credentials owners the following can be seen:

```json
        "owners": [
            {
                "id": 3,
                "type": "organization",
                "name": "Ansible, Inc",
                "description": "Ansible is the simplest way to automate IT.",
                "url": "/api/v2/organizations/3/"
            },
            {
                "id": 3,
                "type": "team",
                "name": "Ansible Users",
                "description": "",
                "url": "/api/v2/teams/3/"
            }
        ]
```

But exporting the data and importing it does not bring the imported credential to the same state. I still not sure why but the imported data is not the same as the one found previously. This is how it is after the import:

```json
        "owners": [
            {
                "id": 2,
                "type": "organization",
                "name": "Ansible, Inc",
                "description": "Ansible is the simplest way to automate IT.",
                "url": "/api/v2/organizations/2/"
            }
        ]
```

We may have some other cases and I still trying to sort all of those out.

jbradberry commented:
regate
